File: haskell2/SC2/Decision.hs
-- Wicked Zerg - Battle Simulation
-- Phase 140: Haskell v2

module BattleSim where

data BattleUnit = BattleUnit
  { unitType :: Int
  , health   :: Float
  , damage   :: Float
  , armor    :: Float
  , posX     :: Float
  , posY     :: Float
  } deriving (Show)

calculateSwarmDamage :: Int -> Int
calculateSwarmDamage count = count * 5

swarmFormation :: Float -> Float -> Int -> Float -> [(Float, Float)]
swarmFormation centerX centerY count radius =
  [ (centerX + radius * cos angle, centerY + radius * sin angle)
  | i <- [0..count-1]
  , let angle = 2 * pi * fromIntegral i / fromIntegral count ]

unitStrength :: Float -> Float -> Float -> Float
unitStrength health damage armor =
  let effective = damage * health / 100
  in effective * (1 - armor * 0.01)

battleOutcome :: [(Float, Float, Float)] -> [(Float, Float, Float)] -> Bool
battleOutcome attackers defenders =
  let attackPower = sum $ map (\(h, d, a) -> unitStrength h d a) attackers
      defensePower = sum $ map (\(h, d, a) -> unitStrength h d a) defenders
  in attackPower > defensePower

File: ocaml2/decision.ml
(* Wicked Zerg - Battle Simulation *)
(* Phase 147: OCaml v2 *)

type battle_unit = {
  unit_type : int;
  health : float;
  damage : float;
  armor : float;
  pos_x : float;
  pos_y : float;
}

let calculate_swarm_damage count =
  count * 5

let swarm_formation center_x center_y count radius =
  let rec aux i acc =
    if i >= count then List.rev acc
    else
      let angle = 2.0 *. Float.pi *. float_of_int i /. float_of_int count in
      let x = center_x +. radius *. cos angle in
      let y = center_y +. radius *. sin angle in
      aux (i + 1) ((x, y) :: acc)
  in
  aux 0 []

let unit_strength health damage armor =
  let effective = damage *. health /. 100.0 in
  effective *. (1.0 -. armor *. 0.01)

let battle_outcome attackers defenders =
  let attack_power = List.fold_left (fun acc (h, d, a) -> 
    acc +. unit_strength h d a) 0.0 attackers in
  let defense_power = List.fold_left (fun acc (h, d, a) ->
    acc +. unit_strength h d a) 0.0 defenders in
  attack_power > defense_power

let () = print_endline "Battle Simulation Initialized - OCaml v2"
